src/GameLogic/Entities/Iceblock.ts

```diff
@@ -0,0 +1,110 @@
+import {Entity} from '../Entity';
+import {EntityType, FloorType} from '../Enums';
+import {Direction8, Direction8Utils} from '../Enums/Direction8';
+import {Level} from '../Level';
+import {Fireball} from './Fireball';
+
+export class Iceblock implements Entity {
+	public readonly type: EntityType;
+
+	public x: number;
+
+	public y: number;
+
+	public direction: Direction8;
+
+	public contains: Entity;
+
+	public melting: boolean;
+
+	public justPushed: boolean;
+
+	constructor(contains: Entity) {
+		this.type = EntityType.Iceblock;
+		this.x = 0;
+		this.y = 0;
+		this.direction = Direction8.None;
+		this.contains = contains;
+		this.melting = contains.type == EntityType.Fireball;
+		this.justPushed = false;
+	}
+
+	public update(level: Level): void {
+		if (this.justPushed) {
+			this.justPushed = false; //Don't update because we've just been moved via push()
+		} else if (this.isMoveAllowed(level, this.direction) && !this.melting) {
+			this.x += Direction8Utils.getX(this.direction);
+			this.y += Direction8Utils.getY(this.direction);
+			this.postMoveProcessing(level);
+		} else {
+			this.direction = Direction8.None;
+		}
+		if (this.melting) {
+			this.contains.x = this.x;
+			this.contains.y = this.y;
+			level.entities.addEntity(this.contains);
+			level.entities.removeEntity(this);
+		}
+	}
+
+	public clone(): Iceblock {
+		const clone = new Iceblock(this.contains);
+		clone.x = this.x;
+		clone.y = this.y;
+		clone.direction = this.direction;
+		clone.melting = this.melting;
+		clone.justPushed = this.justPushed;
+		return clone;
+	}
+
+	public isMoveAllowed(level: Level, direction: Direction8): boolean {
+		const newX = this.x + Direction8Utils.getX(direction);
+		const newY = this.y + Direction8Utils.getY(direction);
+
+		if (newX < 0 || newY < 0 || newX >= level.width || newY >= level.height) {
+			return false;
+		}
+
+		const floor = level.tilesFloor.get(newX, newY);
+
+		if (floor == FloorType.Wall) {
+			return false;
+		}
+
+		const entities = level.entities.getEntitiesAt(newX, newY);
```

Comment:
I think it'd better to block movement by default on any entity, and just handle fireball differently. Whitelist is generally more future proof than a blacklist

Reply:
I'll leave this for now and raise an issue to think about blockability of entities in general and see how we want to handle things like this.